src/action.c

```diff
@@ -1060,7 +1052,24 @@ actions_run(struct view *activator, struct server *server,
 		case ACTION_TYPE_FOCUS_OUTPUT:
 			{
 				const char *output_name = action_get_str(action, "output", NULL);
-				desktop_focus_output(output_from_name(server, output_name));
+				struct output *target = NULL;
+				if (output_name) {
+					target = output_from_name(view->server, output_name);
+				} else {
+					enum view_edge edge = action_get_int(action, "direction", 0);
+					bool wrap = action_get_bool(action, "wrap", false);
+					target = output_get_adjacent(output_nearest_to_cursor(server), edge, wrap);
+				}
+				if (!target) {
+					/*
+					 * Most likely because we're already on the
+					 * output furthest in the requested direction
+					 * or the output or direction was invalid.
+					 */
+					wlr_log(WLR_DEBUG, "Invalid output");
+					break;
+				}
+				desktop_focus_output(target);
```

Comment:
With the change to `output_get_adjacent` suggested below, the entire block for this action (except for the final call to `desktop_focus_output(target)` can be wrapped in a static function in this file, something like
```c
static struct output *
adjacent_output(struct output *output, struct server *server, struct action *action)
{
	const char *output_name = action_get_str(action, "output", NULL);
	struct output *target = NULL;
	if (output_name) {
		target = output_from_name(server, output_name);
	} else {
		enum view_edge edge =
			action_get_int(action, "direction", VIEW_EDGE_INVALID);
		bool wrap = action_get_bool(action, "wrap", false);
		target = output_get_adjacent(output, edge, wrap);
	}

	if (!target) {
		wlr_log(WLR_DEBUG, "Invalid output");
	}

	return target;
}
```

Then, this action can just do
```c
struct output *target = adjacent_output(NULL, server, action);
if (target) {
	desktop_focus_output(target);
}
break;
```
while the `ACTION_TYPE_MOVE_TO_OUTPUT` block can just do
```c
if (!view) {
	break;
}
struct output *target = adjacent_output(NULL, server, action);
if (target) {
	view_move_to_output(view, target);
}
break;
```